zebra-chain/src/primitives/zcash_primitives.rs

```diff
@@ -177,6 +184,92 @@ impl TryFrom<Amount<NonNegative>> for zcash_protocol::value::Zatoshis {
     }
 }
 
+impl TryFrom<Amount> for ZatBalance {
+    type Error = BalanceError;
+
+    fn try_from(amount: Amount) -> Result<Self, Self::Error> {
+        ZatBalance::from_i64(amount.into())
+    }
+}
+
+impl<A: AnchorVariant + Clone> TryFrom<&sapling::ShieldedData<A>>
+    for sapling_crypto::Bundle<Authorized, ZatBalance>
+where
+    sapling::Spend<PerSpendAnchor>: From<(sapling::Spend<A>, A::Shared)>,
+{
+    type Error = Error;
+
+    fn try_from(data: &sapling::ShieldedData<A>) -> Result<Self, Self::Error> {
+        let spends: Vec<SpendDescription<Authorized>> = data
+            .spends_per_anchor()
+            .map(TryFrom::try_from)
+            .collect::<Result<_, _>>()?;
+
+        let outputs: Vec<OutputDescription<GrothProofBytes>> = data
+            .outputs()
+            .map(TryFrom::try_from)
+            .collect::<Result<_, _>>()?;
+
+        let vb = ZatBalance::try_from(*data.value_balance())?;
+
+        let auth = Authorized {
+            binding_sig: *data.binding_sig(),
+        };
+
+        sapling_crypto::Bundle::from_parts(spends, outputs, vb, auth).ok_or(Error::Conversion(
+            "failed to convert sapling::ShieldedData to sapling_crypto::Bundle".into(),
+        ))
+    }
+}
+
+impl TryFrom<sapling::Spend<PerSpendAnchor>> for SpendDescription<Authorized> {
+    type Error = Error;
+
+    fn try_from(spend: sapling::Spend<PerSpendAnchor>) -> Result<Self, Self::Error> {
+        let cv = ValueCommitment::from_bytes_not_small_order(&spend.cv().into())
```

Comment:
Unfortunately this and other `from_bytes()` conversions are CPU intensive. This PR might cause a big impact in Zebra performance.

Possible solutions:

- Add methods to `sapling_crypto` to construct these types from existing points. That might be unlikely to get merged, though they may accept it if it's feature-gated
- Change our types to use the  `sapling_crypto` types. Probably the best option in the long run, but it might be a lot of work

Reply:
I wonder if we should just wrap the lrz tx type at this point.

Reply:
We already do this conversion to compute transaction sighashes, we could re-use that for now. I'm updating https://github.com/ZcashFoundation/zebra/pull/9308 to do that for Orchard now.

Reply:
(ah, I didn't notice that Arya had already commented that :rofl: )

> I wonder if we should just wrap the lrz tx type at this point.

That's likely the long-term solution but that would require even more work. It seems we can achieve something similar by slowly replacing the "leaf" types with LRZ types